src/hotspot/share/opto/macro.cpp

```diff
@@ -958,7 +958,10 @@ bool PhaseMacroExpand::scalar_replacement(AllocateNode *alloc, GrowableArray <Sa
   if (res != nullptr) { // Could be null when there are no users
     res_type = _igvn.type(res)->isa_oopptr();
   }
-
+  // Bufferes in larval state should not be scalarized.
+  if (alloc->_larval) {
+    return false;
+  }
```

Comment:
Does it need to also add such checks in `InlineTypeNode::make_scalar_in_safepoints` ?

Reply:
With new behavior, a value object in larval state is already allocated over heap and should not be scalarized to InlineTypeNode.

Reply:
Yes, that's true. So can we still generate the `InlineTypeNode` with the larval state buffer as its oops like before? And we can avoid its scalarization by just checking the larval state.

Reply:
I think so. We also need to add the same larval check in https://github.com/openjdk/valhalla/blob/lworld%2Bvector/src/hotspot/share/opto/inlinetypenode.cpp#L365 to avoid the scalarization of the `InlineType`.